proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,1864 @@
+
+- Consuming a value in a datatype exactly once means evaluating it to
+  head normal form and consuming its *linear* fields exactly once.
+
+When pattern matching a linear argument, linear fields are introduced
+as linear variables, and unrestricted fields as unrestricted
+variables:
+
+::
+
+  f :: Foo2 ->. A
+  f (Bar2 x y) = x  -- y is unrestricted, hence does not need to be consumed
+
+An exception to this rule is ``newtype`` declarations in GADT syntax:
+``newtype``-s' argument must be linear (see Interactions_
+below).
+
+Base
+~~~~
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reasons, to prevent linear types from
+interfering with newcomers' understanding of the ``Prelude``, this
+proposal does not modify ``base``. Instead, we expect that users will
+publish new libraries on Hackage including more precisely typed
+``base`` functions. One such library has already started `here
+<https://github.com/tweag/linear-base>`_.
+
+Any linear variant of ``base`` need not redefine any of the data types
+defined in ``base``. This is because like for all other data types,
+constructors of (non-GADT) data types in ``base`` are linear under
+this proposal. Since we get to reuse data types, libraries
+implementing linear variants of ``base`` functions remain compatible
+with ``base`` (e.g. there need not be two ``Maybe`` types, two list
+types etc).
+
+The only function which will need to change is ``($)`` because its
+typing rule is built in the type checker. Ignoring the details about
+levity and higher-rank polymorphism in the typing rule, the type
+``($)`` will be:
+
+::
+
+  ($) :: (a :p-> b) ⊸ a :p-> b
+
+Defining a linear variant of ``base`` is out of scope of this
+proposal. Possible future standardisation of the library content is
+the competence of the Core Libraries Committee (CLC). For expository
+purposes of the next sections, however, we assume that such a library
+will at least define the following data type:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+See the paper for intutions about the ``Unrestricted`` data type.
+
+Multiplicities
+~~~~~~~~~~~~~~
+
+.. _Multiplicities:
+
+So far, we have considered only two multiplicities: ``1`` and ``ω``.
+But the metatheory works with any so-called sup-semi-lattice-ordered
+semi-ring (without a 0) of multiplicities. That is: there is a 1,
+a sum and a product with the usual distributivity laws, a (computable)
+order compatible with the sum and product, such that each pair of
+multiplicities has a (computable) join. Even if there is only two
+multiplicities in this proposal, the proposal is structured to allow
+future extensions.
+
+Here is the definition of sum, product and order for this proposal's
+multiplicities (in Haskell pseudo-syntax):
+
+::
+
+   _ + _ = ω
+
+   1 * x = x
+   x * 1 = 1
+   ω * ω = ω
+
```

Comment:
These rules imply `x * y = y`, surely this is wrong?

Reply:
Right. The second one was supposed to be `x*1=x` (1 is neutral).

- [ ] Fix this.